### PR TITLE
Is-954: reward calculation log

### PR DESCRIPTION
**Pull request by inwonkim**
Add reward calculation debugging

- Write output to specified file, db and log
- using rctool, it is possible to query debugging addresses and change configure of calculation debugging(add/delete address or change outputPath)
- using dbtool, can query calculation debugging output

### CODE REVIEW

File: core/context.go [hunk 1]
```diff
@@ -317,15 +317,19 @@ func (idb *IScoreDB) rollbackAccountDB(blockHeight uint64) error {
...
+	PRep           []*PRep
+	PRepCandidates map[common.Address]*PRepCandidate
+	GV             []*GovernanceVariable
 
 	stats    *Statistics
 	Rollback *Rollback
+
+	calcDebugConf   *CalcDebugConfig
+	calcDebugResult *CalcDebugResult
+	calcDebugDB     db.Database
```
Comment (eunsoo-icon) [inline core/context.go]: Combine calculation debug stuff in Context struct

@ context.go
```golang
type Context struct {
  ...
  calcDebug CalcDebug
}
```

@ calculation_debug.go
```golang
type CalcDebug struct {
  conf *CalcDebugConfig
  result *CalcDebugResult
  db db.Databse
}
```

File: core/calculation_debug.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+
+	for i, calcResult := range ctx.calcDebugResult.Results {
+		if calcResult.Address.Equal(&address) {
+			ctx.calcDebugResult.Results =
+				append(ctx.calcDebugResult.Results[:i], ctx.calcDebugResult.Results[i+1:]...)
+		}
+	}
+}
+
+func resetCalcResults(ctx *Context) {
```
Comment (eunsoo-icon) [inline core/calculation_debug.go]: rename to resetCalcDebugResults

File: core/calculation_debug.go [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+		return
+	}
+	defer f.Close()
+	if _, e := f.Write(data); e != nil {
+		log.Printf("Error while write calculation debug result")
+		return
+	}
+}
+
+func writeCalcDebugOutput(ctx *Context) {
```
Comment (eunsoo-icon) [inline core/calculation_debug.go]: rename to writeCalcDebugResultToDB

File: core/calculation_debug.go [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+	bucket, _ := ctx.calcDebugDB.GetBucket("")
+	b, err := ctx.calcDebugResult.Bytes()
+	if err != nil {
+		log.Print("Error while marshaling calculation debug result")
+		return
+	}
+	bucket.Set(ctx.calcDebugResult.ID(), b)
+}
+
```
Comment (eunsoo-icon) [inline core/calculation_debug.go:382]: add writeCalcDebugResult()
``` golang
func writeCalcDebugResult(ctx *Context) {
  writeCalcDebugResultToDB(ctx)
  writeCalcDebugResultToFile(ctx)
}
```

File: core/calculation_debug.go [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	rewards := make([]*Reward, 0)
+	for _, gv := range ctx.GV {
+		reward := NewReward(gv.BlockHeight)
+		rewards = append(rewards, reward)
+	}
+	result.Rewards = rewards
+	ctx.calcDebugResult.Results = append(ctx.calcDebugResult.Results, result)
+}
+
+func writeResultToFile(ctx *Context) {
```
Comment (eunsoo-icon) [inline core/calculation_debug.go]: renwmae to writeCalcDebugResultToFile()

File: core/context.go [hunk 6]
```diff
@@ -560,8 +566,13 @@ func NewContext(dbPath string, dbType string, dbName string, dbCount int) (*Cont
 		return nil, err
 	}
 
+	initCalcDebugConfig(ctx, debugConfigPath)
+
+	// Open calculation debug result DB
+	ctx.calcDebugDB = db.Open(ctx.DB.info.DBRoot, string(db.GoLevelDBBackend), "calculation_debug")
```
Comment (eunsoo-icon) [inline core/context.go]: Do not open initialize process.

Open at start of the calculation if calculation debugging is enabled.

Close at end of the calculation if calcDebugDB was opened.

Users want to query calculation debug result without icon_rc termination.

**Review by eunsoo-icon — CHANGES_REQUESTED** [review]

File: core/context.go (hunk 6 above)
Comment (eunsoo-icon) [reply core/context.go]: Change DB open/close points for optimization. 
Open and close DB when write the calculation debug result.

File: core/calculation_debug.go [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+	for _, gv := range ctx.GV {
+		reward := NewReward(gv.BlockHeight)
+		rewards = append(rewards, reward)
+	}
+	result.Rewards = rewards
+	ctx.calcDebug.result.Results = append(ctx.calcDebug.result.Results, result)
+}
+
+func WriteCalcDebugResult(ctx *Context) {
+	calcDebugDB := db.Open(ctx.DB.info.DBRoot, string(db.GoLevelDBBackend), "calculation_debug")
```
Comment (eunsoo-icon) [inline core/calculation_debug.go:372]: How about using ctx.calcDebug.conf.Output as the DB path?

File: core/calculation_debug.go [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+	"github.com/icon-project/rewardcalculator/common/db"
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+type CalcDebug struct {
+	conf   *CalcDebugConfig
+	result *CalcDebugResult
+	DB     db.Database
```
Comment (eunsoo-icon) [inline core/calculation_debug.go]: unused variable

**Review by eunsoo-icon — CHANGES_REQUESTED** [review]

**Comment by eunsoo-icon** [comment]
Is it safe to do the following while calculation?
* enable calculation debug when there are debug addresses
* add an address when calculation debug is enabled

File: core/calculation_debug.go (hunk 7 above)
Comment (inwonkim) [reply core/calculation_debug.go:372]: ctx.calcDebug.conf.Output was previously required to output the results to a file, but was deleted as it was no longer needed while storing the results in the DB.

**Review by eunsoo-icon — APPROVED** [review]

**Pull request merged**